File: android/app/src/main/java/ai/thetahealth/mirobody/data/chat/ChatRepository.kt
package ai.thetahealth.mirobody.data.chat

import ai.thetahealth.mirobody.data.chat.dto.ChatAttachment
import ai.thetahealth.mirobody.data.chat.dto.ChatStreamEvent
import ai.thetahealth.mirobody.data.chat.dto.ChatStreamRequest
import ai.thetahealth.mirobody.data.chat.dto.ConversationDetail
import ai.thetahealth.mirobody.data.chat.dto.HistoryDeleteRequest
import ai.thetahealth.mirobody.data.chat.dto.ProviderInfo
import ai.thetahealth.mirobody.data.chat.dto.SessionSummary
import ai.thetahealth.mirobody.data.llm.ChatTurn
import ai.thetahealth.mirobody.data.llm.OnDeviceLlmEngine
import ai.thetahealth.mirobody.data.llm.OnDeviceModelSpec
import ai.thetahealth.mirobody.data.net.ensureOk
import ai.thetahealth.mirobody.data.net.unwrap
import kotlinx.coroutines.flow.Flow

class ChatRepository(
    private val api: ChatApi,
    private val streamClient: ChatStreamClient,
    private val onDeviceEngine: OnDeviceLlmEngine,
) {
    suspend fun listProviders(): List<ProviderInfo> =
        api.listProviders().unwrap().flatMap { g ->
            g.providers.map { ProviderInfo(agent = g.agent, provider = it) }
        }

    suspend fun history(page: Int = 0, pageSize: Int = 20): List<SessionSummary> =
        api.history(page = page, pageSize = pageSize).unwrap().summaries

    suspend fun deleteHistory(sessionId: String) {
        api.deleteHistory(HistoryDeleteRequest(sessionId = sessionId)).ensureOk()
    }

    suspend fun conversation(sessionId: String): ConversationDetail =
        api.conversation(sessionId).unwrap()

    fun chat(
        sessionId: String,
        question: String,
        agent: String,
        provider: String,
        language: String,
        subject: Long = 0,
        attachments: List<ChatAttachment> = emptyList(),
        incognito: Boolean = false,
    ): Flow<ChatStreamEvent> = streamClient.stream(
        ChatStreamRequest(
            question = question,
            sessionId = sessionId,
            agent = agent,
            provider = provider,
            language = language,
            subject = if (subject > 0) subject.toString() else null,
            incognito = incognito,
        ),
        attachments = attachments,
    )

    /**
     * Local, offline turn: routes to the on-device engine ([model]) instead of the
     * server. [history]'s final entry is the new user question; earlier entries are
     * conversation context. Emits the same [ChatStreamEvent] flow the SSE path does.
     */
    fun chatOnDevice(history: List<ChatTurn>, model: OnDeviceModelSpec): Flow<ChatStreamEvent> =
        onDeviceEngine.generate(history, model)

    /** Warm an on-device model so the first turn does not pay for its load. */
    suspend fun preloadOnDevice(model: OnDeviceModelSpec) = onDeviceEngine.preload(model)
}

File: ios/Mirobody/Data/Chat/ChatRepository.swift
import Foundation

/// Chat + history API. Mirrors `data/chat/ChatRepository.kt`.
///
/// Endpoints: `GET /api/providers`, `GET /api/history`, `POST /api/history/delete`,
/// and the SSE `POST /api/chat` (delegated to `ChatStreamClient`).
final class ChatRepository {
    private let api: ApiClient
    private let streamClient: ChatStreamClient
    private let onDeviceEngine: OnDeviceLlmEngine

    init(api: ApiClient, settings: SettingsStore, onDeviceEngine: OnDeviceLlmEngine) {
        self.api = api
        self.streamClient = ChatStreamClient(api: api)
        self.onDeviceEngine = onDeviceEngine
    }

    func listProviders() async throws -> [ProviderInfo] {
        let groups: [ProviderGroup] = try await api.get("/api/providers")
        return groups.flatMap { g in g.providers.map { ProviderInfo(agent: g.agent, provider: $0) } }
    }

    func history(page: Int = 0, pageSize: Int = 20) async throws -> [SessionSummary] {
        let resp: HistoryResponse = try await api.get("/api/history?page=\(page)&page_size=\(pageSize)")
        return resp.summaries
    }

    func deleteHistory(sessionId: String) async throws {
        try await api.postEnsureOk("/api/history/delete", HistoryDeleteRequest(sessionId: sessionId))
    }

    /// GET /api/conversation?id= — the full thread for one saved conversation, to
    /// resume it in the chat view. Mirrors `ChatRepository.conversation` on Android.
    func conversation(sessionId: String) async throws -> ConversationDetail {
        let enc = sessionId.addingPercentEncoding(withAllowedCharacters: .urlQueryValueAllowed) ?? sessionId
        return try await api.get("/api/conversation?id=\(enc)")
    }

    func chat(
        sessionId: String,
        question: String,
        agent: String,
        provider: String,
        language: String,
        subject: Int64 = 0,
        attachments: [ChatAttachment] = [],
        incognito: Bool = false
    ) -> AsyncStream<ChatStreamEvent> {
        streamClient.stream(
            ChatStreamRequest(
                question: question,
                sessionId: sessionId,
                agent: agent,
                provider: provider,
                language: language,
                subject: subject > 0 ? String(subject) : nil,
                incognito: incognito
            ),
            attachments: attachments
        )
    }

    /// Local, offline turn: routes to the on-device engine (`model`) instead of the
    /// server. `history`'s final entry is the new user question. Emits the same
    /// `ChatStreamEvent` stream the SSE path does.
    func chatOnDevice(history: [ChatTurn], model: OnDeviceModelSpec) -> AsyncStream<ChatStreamEvent> {
        onDeviceEngine.generate(history: history, model: model)
    }
}
